### ISSUE TITLE
Move create positions permissions to Manage Container

### ISSUE THREAD

**Issue by campmlc** [ArctosDB/arctos#6783]
Request: It appears that permission to create positions within a container currently requires admin_container role. I am requesting this permission be moved back to manage_container role.

Justification: Per the user roles code table, admin_container role has " Administrative scary-level control of containers; control barcode series Prerequisite: manage_container" and  should be restricted to "highly experienced meticulous user with complete understanding of the container model."  
However, manage_container allows operators to "Create, delete and edit containers. " but operators with this access currently cannot create the positions in those containers because that requires admin_container. I have very experienced grad students who I trust with creating containers, but they cannot create the positions in those containers to make the containers usable, which doesn't make sense. I don't want to have to give them admin_container access, as that is truly scary. But I need them to be able to create usable containers with positions in order to do their job. Hence the request.

Alternatives: Giving lots of people admin_container inappropriately, which is what has been happening. But precisely one person per collection or ideally institution should have that access to prevent serious problems of duplicate series. Forcing the one person in an institution who has admin_container access to be the only person allowed to create positions in containers created by others causes a bottleneck in workflows. If an operator is allowed to delete a container or a specimen record, they should be trusted to create the container or edit the record - that is what "manage" level permissions shoud allow.

**Comment by campmlc** [comment]
Alternative 2: create another level of permissions between "manage" and "admin" that could be assigned to experienced and trained operators. This brings up the previously discussed issue of allowing for finer control over permissions so that different tasks can be explicitly selected by collection managers to match the needs of their collections, and operators with access to multiple collections can have different permisssions on a collection-by-collection basis.

**Comment by mkoo** [comment]
This makes sense to me. I vote for expanding manage_container to include create positions, and out of admin.

**Comment by Jegelewicz** [comment]
I vote for expanding manage_container to include create positions, and out of admin.

**Comment by dustymc** [comment]
1. A role has INSERT on a table or it doesn't. Allowing special access for the intersection of (role, container type) would require something more complex (and so more bug-prone, less secure, more cryptic).
2. Positions are the source of many problems, I don't think a system in which random users are randomly creating positions - a concept which has proven to be difficult to understand - can be a great idea.
3. This is a reversal of https://github.com/ArctosDB/arctos/issues/4084#issuecomment-1045453357
4. [Container: Create Position Bulk](https://arctos.database.museum/tools/bulkCreateContainerPositions.cfm) exists for this, and seems a very simple part of any workflow

A new level is possible, but it would need defined in terms of [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) (and I can't see a place for it, whatever that's worth).

**Comment by Jegelewicz** [comment]
> manage_container allows operators to "Create, delete and edit containers

> A role has INSERT on a table or it doesn't. Allowing special access for the intersection of (role, container type)

Isn't creating positions just creating containers (of type position)? If it is, then it seems like moving it would be possible. If it isn't, what exactly is it and why isn't it sufficiently different from other container roles to satisfy CRUD? 

> admin_container role has " Administrative scary-level control of containers; control barcode series

Where does that say anything about creating positions - how is creating posistions part of admin_container?

I am asking so that I can understand how things are separated now.

> Positions are the source of many problems

Maybe because of #6767 ? People think they have space for things but Arctos has done stuff to their entered values. I haven't seen any public issues with positions for a while now. If people are asking for help outside of Github, then it is hard to know that there are problems....

**Comment by dustymc** [comment]
>Isn't creating positions just creating containers

Yes, that is precisely the point.

> then it seems like moving it would be possible

???????????????????????? admin_container exists to create containers! There is no more fundamental administrative task! Do this as suggested by the documentation and everything else just works! There was an entire round of development culminating in the linked Issue leading here; 2/3s of the complexity involving container roles exists to prevent the need for EXACTLY what's being requested here.

>Arctos has done stuff 

No, this predates the week or so of bug by a decade or two, mostly involving position and things like cabinet position [[ link ]](https://arctos.database.museum/info/ctDocumentation.cfm?table=ctcontainer_type#cabinet_position).

Something clearly isn't meshing, let's chat.

**Comment by Jegelewicz** [comment]
> admin_container exists to create containers

is in direct opposition to

> manage_container allows operators to "Create, delete and edit containers.

So the documentation is incorrect or ?

**Comment by dustymc** [comment]
>documentation is incorrect

YES! Thanks, I'll fix that and doublecheck the roles, I may have missed something. (Which probably isn't functionally important while there's only one UI, but still should not be inconsistent.)

**Comment by campmlc** [comment]
Admin container manages barcode series. This can't be same role as create container. That's how we end up with students getting assigned admin. ?

**Comment by dustymc** [comment]
> can't be same

<img width="1055" alt="Screenshot 2023-10-06 at 10 16 24" src="https://github.com/ArctosDB/arctos/assets/5720791/fa6aa876-8e80-4734-bf1c-a51f437c05c4">

**Comment by campmlc** [comment]
That issue (reference?) occurred because a student in another division created positions in a cryovial, and for whatever reason, this can't be corrected by someone with admin container, so it created a problem that @dusty had to fix.
But now we have the problem that instead of taking on the responsibility of creating all containers themselves, collection managers are just giving higher and higher levels of permissions to students. So the solution didn't work. And things are even worse now, because students given "admin_container" just so they can create freeezer boxes to scan things into now have acccess to "Administrative scary-level control of containers; control barcode series".
These are social issues, but they are caused by people needing to do things they can't find a way to do, and then creating work arounds that cause further problems. In this case, there aren't enough staff to create all containers in advance - so this burden gets shifted to students, who mess up. I don't know what checks and balances we need, but it seems clear that we need to move this particular role out of admin, or we need to create a new role just for managing barcode series.

**Comment by dustymc** [comment]
@Jegelewicz perhaps you can dig the notes of that meeting up, I don't have time at the moment, but the "crazy things where they won't fit" thing was addressed by https://github.com/ArctosDB/arctos/issues/4335 - certainly part of the history, but I don't think there's such a direct relationship.

If the constraint is time, I think @Jegelewicz  or I would be happy to push the buttons, we just need to know something of the procedures.

"collection managers are just giving higher and higher levels of permissions to student" might be a strong argument in favor of (3) (hard-partition containers) on https://github.com/ArctosDB/arctos/issues/6789. I do not see any possible way that justifies taking control possibilities from other collections, however.

**Comment by Jegelewicz** [comment]
> higher and higher levels of permissions to student" might be a strong argument in favor of (3)

I pretty much agree with that because as it is I think there is still a lot of possibility for barcode overlap (due to series not being strictly numerical). We could not allow a 5 character base 36 code that starts with D because at some point you will get DGR01 and I don't thing the "check" we have now makes it realistic to find that - it just occurred to me when I was looking for a series for KSB.

> perhaps you can dig the notes of that meeting up,

I'll go look....

**Comment by Jegelewicz** [comment]
I don't know if any of this answers the question - but the notes are in https://docs.google.com/document/d/1on0bbE7Z21YycA0CDeuyDmky96GW72ZcxpeMga29h2Y/edit?usp=sharing

**Comment by dustymc** [comment]
Thanks/blargh. We should record all meetings.

**Comment by Jegelewicz** [comment]
> We should record all meetings.

Nope - nobody will have the time to watch them to find whatever we are looking for. Comments in the code linking to a Github issue?

**Comment by dustymc** [comment]
>nobody will have the time to watch them to find whatever

I would! I'm pretty sure I remember what happened, there are a bunch of comments that support what I think I remember, yet here we are, doing what from here feels like starting over. If nobody cares what happens in meetings then maybe meetings aren't the best use of time, and if meetings are a good use of time then it seems useful to be able to replay them. (And in case there's any doubt, I think meetings can be a GREAT use of time - I'm pretty sure that one solved a problem as it was described.)

Or maybe this is just a problem with how issues are structured - if meetings can't be persistent (I still don't understand the resistance to recording?) then development/issues shouldn't reference them in any way and need to be a lot more thorough. Maybe that's a good practice anyway, but I suspect it will also result in some people reading less.

The comments in the code (I do try, and GitHub has been a great help) would still be me saying what I think I'm already saying in the issues.

**Comment by Jegelewicz** [comment]
I do think the best path is good writing in the issues about resolutions and what is to be done. Part of me really hates all of the agendas and documents all over the place. Also some things are still "owned by someone else" and not always easy to find. I am happy to help make sure we document decisions - it is just that decisions aren't always clear and sometimes stuff gets done a bit informally.

I don't care if we record meetings - in fact I have a reminder on the next AWG to at least record the "feature" part - but I feel pretty sure that unless they include a tutorial (and sometimes even if they do) nobody is going to watch them and it will be just another source of places for me to look when stuff like this comes up. Too many channels means nobody can ever keep up.

This just means more meetings, but maybe it would be good for you and I to meet up after each AWG to just make sure that any concrete development to-dos are documented in the issues? I can take a stab at that but I think that two heads are probably better than one?

**Comment by dustymc** [comment]
> all over the place.

I definitely have trouble finding things.

>when stuff like this comes up

I suspect I'd have reviewed the recordings less than a half-dozen times if they were readily available, and maybe that's just not worth trying to create and store and manage and....

>meet up after each AWG

Sure, or file the issues (I try but sometimes get distracted even with that - because I'm usually doing what I'm doing now and juggling issues while waiting for other stuff to run) and then meet, or whatever. 

Thanks!

**Comment by campmlc** [comment]
Speaking of meetings, happy to have one for this issue, if that would help clarify the situation. I'm not even sure where this one ended up. I can't tell that anything was resolved, and the problem remains.
